### ISSUE TITLE
Ubuntu Provisioner would make more sense as a CloudInit Provisioner

### ISSUE THREAD

**Issue by imduffy15** [Azure/WALinuxAgent#335]
From what I understand of the codebase the ubuntu provisioner mainly exists in order to support the execution of cloud-init.

Ubuntu isn't the only distribution that supports cloud-init, many things use it including Fedora, Debian, RHEL, CentOS. http://cloudinit.readthedocs.io/en/latest/topics/availability.html

Would it not make more sense to rename it to be a CloudInit provisioner and have it enabled/disabled based of a configuration item rather than the OS type + provision being true or false?

Hope I'm not missing something here, if I am I'd love to hear the rational behind the ubuntu specific provisioner.

**Comment by brendandixon** [comment]
@imduffy15 The availability of cloud-init on multiple platforms has been growing while the use of cloud-init within the Azure Ubuntu distributions has existed for a while. We are looking at converging on cloud-init much more broadly across distributions. As we do so, we'll likely make the cosmetic changes, such as renaming providers, to better convey their purpose.

Thank you for looking at our code!

**Comment by imduffy15** [comment]
Thanks for the info @brendandixon I ended up diving into this further.....

Turns out the implementation on the Cloud-Init side for azure is OS specific. For example https://github.com/number5/cloud-init/blob/master/cloudinit/sources/DataSourceAzure.py#L380 won't work on RHEL as no -m option is available on ssh-keygen.

I'm likely to end up customising the Cloud-Init azure datasource to do the minimum stuff required in order to satisfy Azure (stick in the necessary code for a self.report_event("Provision succeed", is_success=True)) and be able to get rid of walinuxagent completely understanding that support for extensions will be lost.
